pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/ReferralSourceService.java

```diff
@@ -0,0 +1,408 @@
+     */
+    public void run(int operationId) {
+        List<UpdateLog> updateLog = new ArrayList<>();
+
+        // for each participant attempt an update and log results
+        for (var entry: participantDataByPtpId.entrySet()) {
+            String ddpParticipantId = entry.getKey();
+            try {
+                UpdateStatus status = updateReferralSource(ddpParticipantId, entry.getValue(),
+                        getParticipantActivities(ddpParticipantId, esIndex));
+                updateLog.add(new UpdateLog(ddpParticipantId, status.name()));
+            } catch (Exception e) {
+                updateLog.add(new UpdateLog(ddpParticipantId, UpdateStatus.ERROR.name(), e.toString()));
+
+                String msg = String.format("Exception in ReferralSourceService.run for participant %s: %s", ddpParticipantId, e);
+                // many of these exceptions will require investigation, but conservatively we will just log
+                // at error level for those that are definitely concerning
+                if (e instanceof DsmInternalError) {
+                    log.error(msg);
+                    e.printStackTrace();
+                } else {
+                    log.warn(msg);
+                }
+            }
+        }
+
+        // update job log record
+        try {
+            String json = gson.toJson(updateLog);
+            AdminOperationRecord.updateOperationRecord(operationId, AdminOperationRecord.OperationStatus.COMPLETED, json);
+        } catch (Exception e) {
+            log.error("Error writing operation log: {}", e.toString());
+        }
+    }
+
+    private static List<String> getParticipantList(String payload) {
+        ReferralSourceRequest req;
+        try {
+            req = new Gson().fromJson(payload, ReferralSourceRequest.class);
+        } catch (Exception e) {
+            throw new DSMBadRequestException("Invalid request format. Payload: " + payload);
+        }
+        List<String> participants = req.getParticipants();
+        if (participants.isEmpty()) {
+            throw new DSMBadRequestException("Invalid request format. Empty participant list");
+        }
+        return participants;
+    }
+
+    protected UpdateStatus updateReferralSource(String ddpParticipantId, List<ParticipantData> dataList, List<Activities> activities) {
+        if (dataList.isEmpty()) {
+            return UpdateStatus.NO_PARTICIPANT_DATA;
+        }
+        if (activities.isEmpty()) {
+            return UpdateStatus.NO_ACTIVITIES;
+        }
+        List<String> refSources = getReferralSources(activities);
+        if (refSources.isEmpty()) {
+            return UpdateStatus.NO_REFERRAL_SOURCE_IN_ACTIVITY;
+        }
+
+        String refSourceId = convertReferralSources(refSources);
+        if (refSourceId.equals(NA_REF_SOURCE)) {
+            return UpdateStatus.NA_REFERRAL_SOURCE;
+        }
+
+        List<ParticipantData> rgpData = dataList.stream().filter(
```

Comment:
Wondering why we need this loop/filtering ? Any chance of getting NON RGP participants in the request ? 
How is the request generated ? Is it from RGP realm ?

Reply:
I will add a comment.
Each ParticipantData has a `field_type_id` that indicates what's in the `data` field. There can be types other than `RGP_PARTICIPANT_DATA`, but those are not relevant for this operation.